app/routes.py

```diff
@@ -1,2 +1,293 @@
+
+@task_bp.route("", methods=["POST"])
+def post_tasks():
+
+    request_body = request.get_json()
+    
+    if all(keys in request_body for keys in ("title","description","completed_at"))== False:
+        return {
+    "details": "Invalid data"
+                },400
+    else:
+        new_task = Task(title=request_body["title"],
+                            description=request_body["description"],
+                            completed_at=request_body["completed_at"])
+        
+        
+
+        db.session.add(new_task) 
+        db.session.commit() 
+        
+        task = Task.query.get(new_task.task_id)
+        return  {
+            "task": task.to_json()
+        }, 201
+
+
+@task_bp.route("/<task_id>", methods=["GET"])
+def get_task(task_id):
+    task = Task.query.get(task_id)
+    
+    if task is None:
+        return make_response ("",404)
+    else:
+        return {"task":task.to_json()},200
+        
+
+@task_bp.route("/<task_id>", methods=["PUT"])
+def put_task(task_id):    
+    task = Task.query.get(task_id)
+    
+    if task is None:
+        return make_response ("",404)
+    else:
+        form_data = request.get_json()
+        
+        task.title=form_data["title"]
+        task.description=form_data["description"]
+        task.completed_at=form_data["completed_at"]
+        
+        db.session.commit()
+           
+        return {
+                "task": task.to_json()
+            }, 200
+
+@task_bp.route("/<task_id>", methods=["DELETE"])
+def delete_task(task_id):
+    task = Task.query.get(task_id)
+    
+    if task is None:
+        return make_response ("",404)
+    else:
+        
+        db.session.delete(task)
+        
+        db.session.commit()
+        
+        
+        return make_response({
+        "details": f'Task {task.task_id} \"{task.title}\" successfully deleted'
+        }, 200)
+
+@task_bp.route("/<task_id>/mark_incomplete", methods=["PATCH"])
+def mark_incomplete(task_id):
+    task = Task.query.get(task_id)
+    if task is None:
+        return make_response ("",404)
+    
+    task.completed_at=None
+    db.session.commit() 
+    
+    return {
+        "task": task.to_json()
+    }, 200
+
+
+@task_bp.route("/<task_id>/mark_complete", methods=["PATCH"])
+def mark_complete(task_id):
+    task = Task.query.get(task_id)
+    if task is None:
+        return make_response ("",404)
+    
+    task.completed_at=datetime.utcnow()
+    db.session.commit()    
+
+
+    url = f"https://slack.com/api/chat.postMessage?channel=task-notifications&text=Someone just completed the task {task.title}"
+
+    
+    
+    token=os.environ.get("bot_user_token")
+  
+   
+    headers_dict={"Authorization":token}
+    response = requests.request("POST", url, headers=headers_dict)
+    return {
+        "task": {
+            "id": task.task_id,
+            "title": task.title,
+            "description":task.description,
+            "is_complete": task.is_complete()
+        }
+    }, 200
```

Comment:
Great job in getting your slack bot to work. A future refactor can include putting this logic into a helper function.